### PR TITLE
Dispatcher must be a callable

### PR DESCRIPTION
**Pull request by gianarb**
This is a good way to standardize dispatcher
implementation and to improve future implementations

/cc @wdalmut @ezimuel :+1:

### CODE REVIEW

**Comment by fntlnz** [comment]
Interesting approach

File: src/App.php [hunk 1]
```diff
@@ -130,7 +132,7 @@ public function run($request = null, $response = null)
         $httpFlow = $this->getEventManager();
 
         try {
-            $routerInfo = $dispatcher->dispatch($request);
+            $routerInfo = $dispatcher($request);
```
Comment (EmanueleMinotto) [inline src/App.php]: What about `call_user_func($dispatcher, $request)` here?

Penny still requires only PHP 5.4+ and the notation `$dispatcher($request)` will be indipendent only from PHP 7: https://3v4l.org/Rd6qm

File: src/App.php [hunk 2]
```diff
@@ -94,7 +94,9 @@ public function getContainer()
     private function getDispatcher()
     {
         $container = $this->container;
-
+        if (!is_callable($container->get("dispatcher"))) {
+            throw new \RuntimeException("Dispatcher must be a callable");
+        }
```
Comment (samsonasik) [inline src/App.php:99]: we need to have consistency for using double quote or single quote, I prefer single quote

Comment (fntlnz) [inline src/App.php:99]: :+1: for this. Also because single quotes doesn't charge the lexer with extra job.

Comment (EmanueleMinotto) [inline src/App.php:99]: @gianarb https://github.com/pennyphp/penny/pull/75

**Comment by wdalmut** [comment]
:+1: @gianarb can you explain here how to use the router callable?

using callable we can use something like this? (see below)

``` php
"router" => function (/*...*/) {
  // my router
}
```

or 

``` php
"router" => new RouterFactory(/*...*/),

class RouterFactory
{
  public function __invoke(/*...*/) {

  }
}
```

Right?

Which kind of data type the router callable have to return? Any interface or just another callable (the executable action)?

``` php
"router" => function (/*...*/) {
  return function (/*...*/) {
    // Here is my dispatch-able controller
  }
}
```

**Comment by gianarb** [comment]
Yes @wdalmut, right your first two examples already work

At the moment after `$dispatcher($router)` I wait a callable..
Before call it I load controller by Di..
https://github.com/pennyphp/penny/blob/master/src/App.php#L151

File: src/App.php (hunk 1 above)
Comment (samsonasik) [inline src/App.php]: anyway, we now require php 5.5 #76

**Comment by wdalmut** [comment]
Here my **Penny** with a single route and Pimple as a container! (when you fix #77)

``` php
<?php
require __DIR__ . '/../vendor/autoload.php';

use GianArb\Penny\App;
use Interop\Container\Pimple\PimpleInterop;
use Zend\EventManager\EventManager;
use Zend\Diactoros\Response\SapiEmitter;

$container = new PimpleInterop();
$container["event_manager"] = function() {
    return new EventManager();
};
$container["dispatcher"] = function($r) {
    return $r["router"];
};
$container["index"] = new Controller();
$container["router"] = function($r) {
    return function ($request) {
        return [
            null,
            ["index", "index"],
            ["ok"],
        ];
    };
};

$emitter = new SapiEmitter();
$app = new App($container);
$emitter->emit($app->run());

class Controller {
    public function index($request, $response)
    {
        $response->getBody()->write("Here is my skeleton");
        return $response;
    }
}
```

You will see `Here is my skeleton`

File: src/App.php (hunk 1 above)
Comment (EmanueleMinotto) [inline src/App.php]: @samsonasik the problem exists for php until 5.6.13 ([see the 3v4l link](https://3v4l.org/Rd6qm))

**Comment by gianarb** [comment]
Easy! :) 
You can add this into the Penny\Container\PimpeFactory

``` php
$container = new PimpleInterop();
$container["event_manager"] = function() {
    return new EventManager();
};
$container["dispatcher"] = function($r) {
    return $r["router"];
};
$container["index"] = new Controller();
$container["router"] = function($r) {
    return function ($request) {
        return [
            null,
            ["index", "index"],
            ["ok"],
        ];
    };
};
```

To reuse it :D

File: src/App.php (hunk 1 above)
Comment (samsonasik) [inline src/App.php]: yup, seems `call_user_func` is for the rescue now

Comment (samsonasik) [inline src/App.php]: @gianarb  I created PR #78 for it

**Pull request merged**